demos/mqtt/mqtt_demo_basic_tls/mqtt_demo_basic_tls.c

```diff
@@ -289,108 +610,252 @@ static int32_t transportRecv( MQTTNetworkContext_t pSslContext,
+    }
+    else
+    {
+        LogError( ( "Poll returned with status = %d.", pollStatus ) );
+        bytesReceived = -1;
+    }
+
+    return bytesReceived;
+}
+
+/*-----------------------------------------------------------*/
+
+static uint32_t getTimeMs( void )
+{
+    uint32_t timeMs;
+    struct timespec timeSpec;
+
+    /* Get the MONOTONIC time. */
+    clock_gettime( CLOCK_MONOTONIC, &timeSpec );
+
+    /* Calculate the milliseconds from timespec. */
+    timeMs = ( uint32_t ) ( timeSpec.tv_sec * 1000 )
+             + ( uint32_t ) ( timeSpec.tv_nsec / ( 1000 * 1000 ) );
+
+    /* Reduce globalEntryTime from obtained time so as to always return the
+     * elapsed time in the application. */
+    timeMs = ( uint32_t ) ( timeMs - globalEntryTimeMs );
+
+    return timeMs;
 }
 
 /*-----------------------------------------------------------*/
 
-/**
- * @brief The timer query function provided to the MQTT context.
- *
- * Currently not implemented.
- */
-static uint32_t getTime( void )
+static void handleIncomingPublish( MQTTPublishInfo_t * pPublishInfo,
+                                   uint16_t packetIdentifier )
 {
-    return 0;
+    assert( pPublishInfo != NULL );
+
+    /* Process incoming Publish. */
+    LogInfo( ( "Incoming QOS : %d.", pPublishInfo->qos ) );
+
+    /* Verify the received publish is for the we have subscribed to. */
+    if( ( pPublishInfo->topicNameLength == MQTT_EXAMPLE_TOPIC_LENGTH ) &&
+        ( 0 == strncmp( MQTT_EXAMPLE_TOPIC,
+                        pPublishInfo->pTopicName,
+                        pPublishInfo->topicNameLength ) ) )
+    {
+        LogInfo( ( "Incoming Publish Topic Name: %.*s matches subscribed topic.",
+                   pPublishInfo->topicNameLength,
+                   pPublishInfo->pTopicName ) );
+        LogInfo( ( "Incoming Publish Message : %.*s.\n\n",
+                   ( int ) pPublishInfo->payloadLength,
+                   ( const char * ) pPublishInfo->pPayload ) );
+    }
+    else
+    {
+        LogInfo( ( "Incoming Publish Topic Name: %.*s does not match subscribed topic.",
+                   pPublishInfo->topicNameLength,
+                   pPublishInfo->pTopicName ) );
+    }
 }
 
 /*-----------------------------------------------------------*/
 
-/**
- * @brief MQTT event callback.
- *
- * Currently not implemented.
- */
 static void eventCallback( MQTTContext_t * pContext,
                            MQTTPacketInfo_t * pPacketInfo,
                            uint16_t packetIdentifier,
                            MQTTPublishInfo_t * pPublishInfo )
 {
+    assert( pContext != NULL );
+    assert( pPacketInfo != NULL );
+
+    /* Handle incoming publish. The lower 4 bits of the publish packet
+     * type is used for the dup, QoS, and retain flags. Hence masking
+     * out the lower bits to check if the packet is publish. */
+    if( ( pPacketInfo->type & 0xF0U ) == MQTT_PACKET_TYPE_PUBLISH )
```

Comment:
This could just be moved to the switch since SUBACK, UNSUBACK, and PINGRESP all have their lower four bits set to 0

Reply:
I thought about that and intentionally made the change this way to make it clear to the reader that only PUBLISH packet type will have to mask the lower bits.